### PR TITLE
chore: remove test sleeps, fix flaky test

### PR DESCRIPTION
**Pull request by toddbaert**
This PR fixes a flaky events test and removes a bunch of sleeps with a new `AssertUntilAsync` util.

It brings the test suite time down to 5s from 15s on my machine.

### CODE REVIEW

File: test/OpenFeature.Tests/OpenFeatureEventTests.cs [hunk 1]
```diff
@@ -257,8 +256,12 @@ public async Task API_Level_Event_Handlers_Should_Be_Executed_When_Other_Handler
             var testProvider = new TestProvider(fixture.Create<string>());
             await Api.Instance.SetProvider(testProvider);
 
-            failingEventHandler.Received().Invoke(Arg.Is<ProviderEventPayload>(payload => payload.ProviderName == testProvider.GetMetadata().Name));
-            eventHandler.Received().Invoke(Arg.Is<ProviderEventPayload>(payload => payload.ProviderName == testProvider.GetMetadata().Name));
```
Comment (toddbaert) [inline test/OpenFeature.Tests/OpenFeatureEventTests.cs:261]: This was the flaky test (in 10 runs it failed twice locally). Notice it didn't have a sleep. I believe there was a race condition where only 1 of these 2 handlers had run by the time we hit the assertions.

File: test/OpenFeature.Tests/TestUtils.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+internal class Utils
+{
+    /// <summary>
+    /// Repeatedly runs the supplied assertion until it doesn't throw, or the timeout is reached.
+    /// </summary>
+    /// <param name="assertionFunc">Function which makes an assertion</param>
+    /// <param name="timeoutMillis">Timeout in millis (defaults to 1000)</param>
+    /// <param name="pollIntervalMillis">Poll interval (defaults to 100</param>
+    /// <returns></returns>
+    public static async Task AssertUntilAsync(Action assertionFunc, int timeoutMillis = 1000, int pollIntervalMillis = 100)
```
Comment (toddbaert) [inline test/OpenFeature.Tests/TestUtils.cs:14]: Java has "Awaitilty", which is a nice lib for this, I could find no C# equivalent (neither could @bacherfl IIRC).

**Review by austindrenski — CHANGES_REQUESTED** [review]
Suggestion for the `AssertUntilAsync(...)` impl., but otherwise looks good!

File: test/OpenFeature.Tests/TestUtils.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                lastException = e;
+            }
+            finally
+            {
+                await Task.Delay(pollIntervalMillis).ConfigureAwait(false);
+            }
+        }
+        while (stopwatch.Elapsed.TotalMilliseconds < timeoutMillis);
+        throw lastException;
+    }
```
[the remark this answers is not in the document]
Comment (toddbaert) [reply test/OpenFeature.Tests/TestUtils.cs:38]: Love it, all great ideas, and all well-explained!

I am accepting, but I had to pass a default token to `CreateLinkedTokenSource`, and `using declarations` aren't valid in <8, so I used a traditional block.

File: test/OpenFeature.Tests/TestUtilsTest.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+using Xunit;
+
+namespace OpenFeature.Tests
+{
+    public class TestUtilsTest
+    {
+        [Fact]
+        public async void Should_Fail_If_Assertion_Fails()
+        {
+            await Assert.ThrowsAnyAsync<Exception>(() => Utils.AssertUntilAsync(() => Assert.True(1.Equals(2)), 100, 10)).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (toddbaert) [reply test/OpenFeature.Tests/TestUtilsTest.cs:14]: absorbed this everywhere it was applicable.

File: test/OpenFeature.Tests/TestUtils.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (austindrenski) [reply test/OpenFeature.Tests/TestUtils.cs]: haha, I was just about to comment that, but you beat me to it!

It's such a sharp corner, and so easy to stub your toes on. I'm surprised they still haven't shipped an overload to say `please wait this long, unless cancelled, in which case return without throwing`, but alas, maybe in `net9.0`...

File: test/OpenFeature.Tests/TestUtils.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                try
+                {
+                    await Task.Delay(pollIntervalMillis, cts.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    throw new AggregateException(message, exceptions);
+                }
+            }
+            throw new AggregateException(message, exceptions);
```
Comment (austindrenski) [inline test/OpenFeature.Tests/TestUtils.cs:51]: Also nice catch on the race condition during the jump from `L44` back to `L26`! ...now to go fix a bunch of internal code that is most definitely susceptible to this..........

**Review by austindrenski — APPROVED** [review]
LGTM!

**Review by beeme1mr — APPROVED** [review]

File: test/OpenFeature.Tests/TestUtils.cs (hunk 5 above)
Comment (toddbaert) [reply test/OpenFeature.Tests/TestUtils.cs:51]: :sweat_smile:

**Pull request merged**